HWs/HW3/mostCommon.c

```diff
@@ -0,0 +1,236 @@
+    {
+        int j = i + 1;
+
+        while (j > leftIndex && array[j - 1] > array[j])
+        {
+            swap(&array[j - 1], &array[j]);
+            --j;
+        }
+    }
+}
+
+void quickSort(int array[], int leftIndex, int rightIndex)
+{
+    if (rightIndex - leftIndex <= 8)
+    {
+        insertsSort(array, leftIndex, rightIndex);
+    }
+    else
+    {
+        int reference = partition(array, leftIndex, rightIndex);
+        quickSort(array, leftIndex, reference - 1);
+        quickSort(array, reference + 1, rightIndex);
+    }  
+}
+
+void mostCommon(int array[], int length, int *mostCommon, int *counterMaximum)
+{
+    int counter = 1;
+
+    for (int i = 1; i < length; ++i)
+    {
+        if (array[i] == array[i - 1])
+        {
+            ++counter;
+        }
+        else 
+        {
+            if (counter > *counterMaximum)
+            {
+                *mostCommon = array[i - 1];
+                *counterMaximum = counter;
+            }
+
+            counter = 1;
+        }
+    }
+
+    if (counter > *counterMaximum)  // If the most common number is at the end, then we never check the last element for the maximum
+    {
+        *mostCommon = array[length - 2]; 
+        *counterMaximum = counter;
+    }
+}
+
+bool testQuickSort1()
+{
+    int array[5] = {5, 4, 3, 2, 1};
+
+    quickSort(array, 0, 4);
+
+    for (int i = 0; i < 5; ++i)
+    {
+        if (array[i] != i + 1)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool testQuickSort2()
+{
+    int array[4] = {-3, 0, -1, -2};
+    int correctArray[4] = {-3, -2, -1, 0};
+
+    quickSort(array, 0, 3);
+
+    for (int i = 0; i < 4; ++i)
+    {
+        if (array[i] != correctArray[i])
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool testQuickSort3()
+{
+    int array[6] = {2, 2, 2, 2, 2, 2};
+
+    quickSort(array, 0, 5);
+
+    for (int i = 0; i < 6; ++i)
+    {
+        if (array[i] != 2)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+int testQuickSort()
+{
+    if (!testQuickSort1())
+    {
+        return -1;
+    }
+    
+    if (!testQuickSort2())
+    {
+        return -2;
+    }
+
+    if (!testQuickSort3())
+    {
+        return -3;
+    }
+
+    return 0;
+}
+
+bool testMostCommon()
+{
+    int array[5] = {0, 1, 1, 1, 0};
+    int counterMaximum = 1;
+    int mostCommonNumber = array[0];
+    mostCommon(array, 5, &mostCommonNumber, &counterMaximum);
+
+    return mostCommonNumber == 1 && counterMaximum == 3;
+}
+
+int main()
+{
+    int errorCode = testQuickSort();
+    if (errorCode != 0)
+    {
+        printf("QuickSort function tests failed! Error code %d.\n", errorCode);
+        return 0;
+    }
+
+    if (!testMostCommon())
+    {
+        printf("MostCommon function tests failed!\n");
+        return 0;
+    }
+
+    int array[arraySizeBound] = { 0 };
+    int arraySize = 0;
+    int scan_res = 0;
```

Comment:
```suggestion
    int scanRes = 0;
```
Единообразие (а сначала опечатался, и получилось слово "единообразик" =)) )